patroni/ctl.py

```diff
@@ -1372,3 +1373,246 @@ def format_pg_version(version):
+    - ``postgresql.listen``: source instance's listen_addresses and port GUC values
+    - ``postgresql.connect_address``: if possible, generated from the connection params
+    - ``postgresql.authentication``:
+        - superuser and replication users defined (if possible, usernames are set from the respective Patroni ENV vars,
+          otherwise the default 'postgres' and 'replicator' values are used).
+          If not a sample config, either DSN or PG ENV vars are used to define superuser authentication parameters.
+        - rewind user is defined for a sample config if PG version can be defined and PG version is 11+
+          (if possible, username is set from the respective Patroni ENV var)
+    - ``bootsrtap.dcs.postgresql.use_pg_rewind set to True if PG version is 11+
+    - ``postgresql.pg_hba`` defaults or the lines gathered from the source instance's hba_file
+    - ``postgresql.pg_ident`` the lines gathered from the source instance's ident_file
+
+    :param scope: Scope parameter value to write into the configuration.
+    :param file: Full path to the configuration file to be created (/tmp/patroni.yml by default).
+    :param sample: Optional flag. If set, no source instance will be used - generate config with some sane defaults.
+    :param dsn: Optional DSN string for the local instance to get GUC values from.
+    :param bin_dir: Optional path to Postgres binaries.
+    """
+    from patroni.config import Config
+    from patroni.validator import get_major_version
+
+    no_value_msg = '#FIXME'
+    pg_version = None
+
+    dynamic_config = Config.get_default_config()
+    dynamic_config['postgresql']['parameters'] = dict(dynamic_config['postgresql']['parameters'])
+    config = {
+        'scope': scope or no_value_msg,
+        'name': socket.gethostname(),
+        'bootstrap': {
+            'dcs': dynamic_config
+        },
+        'postgresql': {
+            'parameters': None,
+            'connect_address': no_value_msg,
+            'listen': no_value_msg,
+        },
+    }
+    # for a running instance bin_dir will be gathered from the postmaster process
+    if sample and bin_dir:
+        config['postgresql']['bin_dir'] = bin_dir
+
+    if sample:  # some sane defaults or values set via Patroni env vars
+        replicator = os.getenv('PATRONI_REPLICATION_USERNAME', 'replicator')
+        config['postgresql']['authentication'] = {
+            'superuser': {
+                'username': os.getenv('PATRONI_SUPERUSER_USERNAME', 'postgres'),
+                'password': os.getenv('PATRONI_SUPERUSER_PASSWORD', no_value_msg)
+            },
+            'replication': {
+                'username': replicator,
+                'password': os.getenv('PATRONI_REPLICATION_PASSWORD', no_value_msg)
+            }
+        }
+        config['postgresql']['pg_hba'] = [
+            'host all all 0.0.0.0/0 md5',
+            f'host replication {replicator} all md5'
```

Comment:
Should we in case of `sample` config insist on using `scram-sha-256` starting from a certain version?